Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayerSkia.cpp

```diff
@@ -28,90 +28,27 @@
-    paintGraphicsLayerContents(context, clipRect);
-}
-
 Ref<CoordinatedTileBuffer> CoordinatedGraphicsLayer::paintTile(const IntRect& dirtyRect)
 {
-    auto paintBuffer = [&](CoordinatedTileBuffer& buffer) {
-        buffer.beginPainting();
-
-        if (auto* canvas = buffer.canvas()) {
-            canvas->save();
-            canvas->clear(SkColors::kTransparent);
-
-            GraphicsContextSkia context(*canvas, buffer.isBackedByOpenGL() ? RenderingMode::Accelerated : RenderingMode::Unaccelerated, RenderingPurpose::LayerBacking);
-            paintIntoGraphicsContext(context, dirtyRect);
-
-            canvas->restore();
-        }
-
-        buffer.completePainting();
-    };
-
-    // Skia/GPU - accelerated rendering.
-    auto* skiaGLContext = PlatformDisplay::sharedDisplay().skiaGLContext();
-    if (auto* acceleratedBitmapTexturePool = m_coordinator->skiaAcceleratedBitmapTexturePool(); acceleratedBitmapTexturePool && skiaGLContext->makeContextCurrent()) {
-        OptionSet<BitmapTexture::Flags> textureFlags;
-        if (!contentsOpaque())
-            textureFlags.add(BitmapTexture::Flags::SupportsAlpha);
-
-        auto buffer = CoordinatedAcceleratedTileBuffer::create(acceleratedBitmapTexturePool->acquireTexture(dirtyRect.size(), textureFlags));
-        WTFBeginSignpost(this, PaintTile, "Skia accelerated, dirty region %ix%i+%i+%i", dirtyRect.x(), dirtyRect.y(), dirtyRect.width(), dirtyRect.height());
-        paintBuffer(buffer.get());
-        WTFEndSignpost(this, PaintTile);
+    auto& paintingEngine = m_coordinator->skiaPaintingEngine();
 
-        return buffer;
-    }
+    // ### Asynchronous rendering on worker threads ###
+    if (paintingEngine.canPerformThreadedAcceleratedRendering())
+        return paintingEngine.postPaintingTask(RenderingMode::Accelerated, *this, dirtyRect);
 
-    auto buffer = CoordinatedUnacceleratedTileBuffer::create(dirtyRect.size(), contentsOpaque() ? CoordinatedTileBuffer::NoFlags : CoordinatedTileBuffer::SupportsAlpha);
+    if (paintingEngine.canPerformThreadedUnacceleratedRendering())
+        return paintingEngine.postPaintingTask(RenderingMode::Unaccelerated, *this, dirtyRect);
 
-    // Skia/CPU - threaded unaccelerated rendering.
-    if (auto* workerPool = m_coordinator->skiaThreadedPaintingPool()) {
-        workerPool->postPaintingTask(buffer, *this, dirtyRect);
-        return buffer;
-    }
+    // ### Synchronous rendering on main thread ###
+    if (paintingEngine.canPerformAcceleratedRendering())
+        return paintingEngine.performPaintingTask(RenderingMode::Accelerated, *this, dirtyRect);
 
-    // Blocking, single-thread variant.
-    WTFBeginSignpost(this, PaintTile, "Skia unaccelerated, dirty region %ix%i+%i+%i", dirtyRect.x(), dirtyRect.y(), dirtyRect.width(), dirtyRect.height());
-    paintBuffer(buffer.get());
-    WTFEndSignpost(this, PaintTile);
-    return buffer;
+    return paintingEngine.performPaintingTask(RenderingMode::Unaccelerated, *this, dirtyRect);
```

Comment:
So, in the end we are always calling the same function but with different rendering mode, so I would add helper function or lambda, and just call `paintingEngine.performPaintingTask(renderingMode(), *this, dirtyRect);`

Reply:
Hmm, what should `renderingMode()` be? A `CoordinatedGraphicsLayer` has no rendering mode per-se.

Maybe it's a bit hard to see in the diff, but we are first determining the rendering mode, by checking conditions in `SkiaPaintingEngine`, before we call `performPaintingTask`:

```
Ref<CoordinatedTileBuffer> CoordinatedGraphicsLayer::paintTile(const IntRect& dirtyRect)
{
    auto& paintingEngine = m_coordinator->skiaPaintingEngine();

    // ### Asynchronous rendering on worker threads ###
    if (paintingEngine.canPerformThreadedAcceleratedRendering())
        return paintingEngine.postPaintingTask(RenderingMode::Accelerated, *this, dirtyRect);

    if (paintingEngine.canPerformThreadedUnacceleratedRendering())
        return paintingEngine.postPaintingTask(RenderingMode::Unaccelerated, *this, dirtyRect);

    // ### Synchronous rendering on main thread ###
    if (paintingEngine.canPerformAcceleratedRendering())
        return paintingEngine.performPaintingTask(RenderingMode::Accelerated, *this, dirtyRect);

    return paintingEngine.performPaintingTask(RenderingMode::Unaccelerated, *this, dirtyRect);
}
```

And we also do it in a specific order. First mt-GPU, then mt-CPU, then GPU, then CPU.

But  you are right, that we could just call: `paintingEngine.performPaintingTask(*this, dirtyRect)` (without a rendering mode specified, and let `SkiaPaintingEngine` figure it out, based on the same logic).

Reply:
I reworked the logic, instead of the three bool accessors to determine the painting mode (threaded/main thread, CPU/GPU) I came up with this:

```
static bool canPerformAcceleratedRendering()
{
    return ProcessCapabilities::canUseAcceleratedBuffers() && PlatformDisplay::sharedDisplay().skiaGLContext();
}

RenderingMode SkiaPaintingEngine::renderingMode() const
{
    if (canPerformAcceleratedRendering())
        return RenderingMode::Accelerated;

    return RenderingMode::Unaccelerated;
}

std::optional<RenderingMode> SkiaPaintingEngine::threadedRenderingMode() const
{
    if (m_gpuWorkerPool && canPerformAcceleratedRendering())
        return RenderingMode::Accelerated;

    if (m_cpuWorkerPool)
        return RenderingMode::Unaccelerated;

    return std::nullopt;
}
```

and a new `paintLayer` method:

```
Ref<CoordinatedTileBuffer> SkiaPaintingEngine::paintLayer(const CoordinatedGraphicsLayer& layer, const IntRect& dirtyRect)
{
    // ### Asynchronous rendering on worker threads ###
    if (auto renderingMode = SkiaPaintingEngine::threadedRenderingMode())
        return postPaintingTask(renderingMode.value(), layer, dirtyRect);

    // ### Synchronous rendering on main thread ###
    return performPaintingTask(renderingMode(), layer, dirtyRect);
}
```

This makes `paintTile` trivial:

```
Ref<CoordinatedTileBuffer> CoordinatedGraphicsLayer::paintTile(const IntRect& dirtyRect)
{
    auto& paintingEngine = m_coordinator->skiaPaintingEngine();
    return paintingEngine.paintLayer(*this, dirtyRect);
}
```

Reply:
> Hmm, what should `renderingMode()` be? A `CoordinatedGraphicsLayer` has no rendering mode per-se.

A simple static helper or a lambda. 

> Maybe it's a bit hard to see in the diff, but we are first determining the rendering mode, by checking conditions in `SkiaPaintingEngine`, before we call `performPaintingTask`:
> 
> ```
> Ref<CoordinatedTileBuffer> CoordinatedGraphicsLayer::paintTile(const IntRect& dirtyRect)
> {
>     auto& paintingEngine = m_coordinator->skiaPaintingEngine();
> 
>     // ### Asynchronous rendering on worker threads ###
>     if (paintingEngine.canPerformThreadedAcceleratedRendering())
>         return paintingEngine.postPaintingTask(RenderingMode::Accelerated, *this, dirtyRect);
> 
>     if (paintingEngine.canPerformThreadedUnacceleratedRendering())
>         return paintingEngine.postPaintingTask(RenderingMode::Unaccelerated, *this, dirtyRect);
> 
>     // ### Synchronous rendering on main thread ###
>     if (paintingEngine.canPerformAcceleratedRendering())
>         return paintingEngine.performPaintingTask(RenderingMode::Accelerated, *this, dirtyRect);
> 
>     return paintingEngine.performPaintingTask(RenderingMode::Unaccelerated, *this, dirtyRect);
> }
> ```

So, the helper would be exactly the same, but instead of calling the paint method, it would return the rendering mode. What I didn't notice in the diff is that we are not actually calling the same paint method, but post/perform, so it's not really possible, sorry for the noise.

> And we also do it in a specific order. First mt-GPU, then mt-CPU, then GPU, then CPU.
> 
> But you are right, that we could just call: `paintingEngine.performPaintingTask(*this, dirtyRect)` (without a rendering mode specified, and let `SkiaPaintingEngine` figure it out, based on the same logic).

That would make paintTile simpler indeed.